modules/apps/callbacks/types/callbacks_test.go

```diff
@@ -0,0 +1,589 @@
+		})
+	}
+}
+
+func (s *CallbacksTypesTestSuite) TestUserDefinedGasLimit() {
+	denom := ibctesting.TestCoin.Denom
+	amount := ibctesting.TestCoin.Amount.String()
+	sender := sdk.AccAddress(secp256k1.GenPrivKey().PubKey().Address()).String()
+	receiver := sdk.AccAddress(secp256k1.GenPrivKey().PubKey().Address()).String()
+
+	testCases := []struct {
+		name       string
+		packetData transfertypes.FungibleTokenPacketData
+		expUserGas uint64
+	}{
+		{
+			"success: memo is empty",
+			transfertypes.FungibleTokenPacketData{
+				Denom:    denom,
+				Amount:   amount,
+				Sender:   sender,
+				Receiver: receiver,
+				Memo:     "",
+			},
+			0,
+		},
+		{
+			"success: memo has user defined gas limit",
+			transfertypes.FungibleTokenPacketData{
+				Denom:    denom,
+				Amount:   amount,
+				Sender:   sender,
+				Receiver: receiver,
+				Memo:     `{"src_callback": {"gas_limit": "100"}}`,
+			},
+			100,
+		},
+		{
+			"failure: memo has empty src_callback object",
+			transfertypes.FungibleTokenPacketData{
+				Denom:    denom,
+				Amount:   amount,
+				Sender:   sender,
+				Receiver: receiver,
+				Memo:     `{"src_callback": {}}`,
+			},
+			0,
+		},
+		{
+			"failure: memo has user defined gas limit as json number",
+			transfertypes.FungibleTokenPacketData{
+				Denom:    denom,
+				Amount:   amount,
+				Sender:   sender,
+				Receiver: receiver,
+				Memo:     `{"src_callback": {"gas_limit": 100}}`,
+			},
+			0,
+		},
+		{
+			"failure: memo has user defined gas limit as negative",
+			transfertypes.FungibleTokenPacketData{
+				Denom:    denom,
+				Amount:   amount,
+				Sender:   sender,
+				Receiver: receiver,
+				Memo:     `{"src_callback": {"gas_limit": "-100"}}`,
+			},
+			0,
+		},
+		{
+			"failure: memo has user defined gas limit as string",
+			transfertypes.FungibleTokenPacketData{
+				Denom:    denom,
+				Amount:   amount,
+				Sender:   sender,
+				Receiver: receiver,
+				Memo:     `{"src_callback": {"gas_limit": "invalid"}}`,
+			},
+			0,
+		},
+		{
+			"failure: memo has user defined gas limit as empty string",
+			transfertypes.FungibleTokenPacketData{
+				Denom:    denom,
+				Amount:   amount,
+				Sender:   sender,
+				Receiver: receiver,
+				Memo:     `{"src_callback": {"gas_limit": ""}}`,
+			},
+			0,
+		},
+		{
+			"failure: malformed memo",
+			transfertypes.FungibleTokenPacketData{
+				Denom:    denom,
+				Amount:   amount,
+				Sender:   sender,
+				Receiver: receiver,
+				Memo:     `invalid`,
+			},
+			0,
+		},
+	}
+
+	for _, tc := range testCases {
+		callbackData, ok := tc.packetData.GetCustomPacketData(types.SourceCallbackKey).(map[string]interface{})
+		s.Require().Equal(ok, callbackData != nil)
+		s.Require().Equal(tc.expUserGas, types.GetUserDefinedGasLimit(callbackData), tc.name)
+	}
+}
+
+func (s *CallbacksTypesTestSuite) TestGetCallbackDataErrors() {
+	// Success cases are tested above. This test case tests extra error case where
```

Comment:
would additionally be nice to explicitly check for the error return for `ErrCallbackKeyNotFound`